packages/labs/src/tagInput/tagInput.tsx

```diff
@@ -0,0 +1,137 @@
+     * Callback invoked when a new tag is added by the user pressing `enter` on the input.
+     * Receives the current value of the input field. New tags are expected to be appended to
+     * the list.
+     */
+    onAdd?: (value: string) => void;
+
+    /**
+     * Callback invoked when the user clicks the X button on a tag.
+     * Receives value and index of removed tag.
+     */
+    onRemove?: (value: string, index: number) => void;
+
+    /**
+     * React props to pass to each `Tag`. Provide an object to pass the same props to every tag,
+     * or a function to customize props per tag.
+     *
+     * If you define `onRemove` here then you will have to implement your own tag removal
+     * handling as the `TagInput` `onRemove` prop will never be invoked.
+     */
+    tagProps?: ITagProps | ((value: string, index: number) => ITagProps);
+
+    /** Controlled tag values. */
+    values: string[];
+}
+
+export interface ITagInputState {
+    inputValue?: string;
+    isInputFocused?: boolean;
+}
+
+@PureRender
+export class TagInput extends AbstractComponent<ITagInputProps, ITagInputState> {
+    public static displayName = "Blueprint.TagInput";
+
+    public static defaultProps: Partial<ITagInputProps> = {
+        inputProps: {},
+        tagProps: {},
+    };
+
+    public state: ITagInputState = { inputValue: "", isInputFocused: false };
+
+    private inputElement: HTMLInputElement;
+    private refHandlers = {
+        input: (ref: HTMLInputElement) => {
+            this.inputElement = ref;
+            const refHandler = this.props.inputProps.ref;
+            if (Utils.isFunction(refHandler)) {
+                refHandler(ref);
+            }
+        },
+    };
+
+    public render() {
+        const { inputProps, tagProps, values } = this.props;
+
+        const classes = classNames(Classes.INPUT, "pt-tag-input", {
+            [Classes.ACTIVE]: this.state.isInputFocused,
+        }, this.props.className);
+
+        const tags = values.map((tag, i) => {
+            const props = Utils.isFunction(tagProps) ? tagProps(tag, i) : tagProps;
+            return <Tag data-tag-index={i} key={tag + "__" + i} onRemove={this.handleRemoveTag} {...props}>{tag}</Tag>;
+        });
+        return (
+            <div className={classes} onClick={this.handleContainerClick}>
+                {tags}
+                <input
+                    value={this.state.inputValue}
+                    {...inputProps}
+                    onBlur={this.handleInputBlur}
+                    onFocus={this.handleInputFocus}
+                    onChange={this.handleInputChange}
+                    onKeyDown={this.handleInputKeyDown}
+                    ref={this.refHandlers.input}
+                    className={classNames("pt-input-invisible", inputProps.className)}
+                />
+            </div>
+        );
+    }
+
+    private handleContainerClick = () => {
+        if (this.inputElement != null) {
+            this.inputElement.focus();
```

Comment:
If I'm blurred and then click a tag's remove button, is it still expected that the input focus?